File: sdk/src/workflows/internal/csv.ts
export type CsvSafety = "spreadsheet" | "raw";

type JsonRecord = Record<string, unknown>;

type FieldDescriptor = {
  identity: string;
  key?: string | undefined;
  label: string;
  missingIndex?: number | undefined;
  column?: string | undefined;
};

export type CsvSchema = {
  topKeys: string[];
  descriptors: FieldDescriptor[];
};

/**
 * Build a deterministic CSV schema from one page and the board's field
 * definitions. The seed page is intentionally bounded; later pages cannot
 * change the header or cause an unbounded discovery pass.
 */
export function createItemsCsvSchema(items: JsonRecord[], fieldDefinitions: JsonRecord[] = []): CsvSchema {
  const topKeys = Array.from(new Set(items.flatMap((item) => Object.keys(item).filter((key) => key !== "fields")))).sort(compareText);
  const descriptors = new Map<string, FieldDescriptor>();
  let missingOccurrence = 0;

  for (const definition of fieldDefinitions) {
    const label = fieldLabel(definition);
    if (label === "") continue;
    const key = typeof definition["key"] === "string" && definition["key"] !== "" ? definition["key"] : undefined;
    const identity = key === undefined ? `missing:${label}:${++missingOccurrence}` : `key:${key}`;
    if (!descriptors.has(identity)) descriptors.set(identity, { identity, key, label });
  }
  for (const item of items) collectFieldValues(item, descriptors);
  return finalizeCsvSchema(topKeys, descriptors);
}

/** Render one page against a previously frozen CSV schema. */
export function renderItemsCsvChunk(items: JsonRecord[], safety: CsvSafety, schema: CsvSchema, includeHeader: boolean): string {
  if (items.length === 0 && !includeHeader) return "";
  const fieldValues = items.map((item) => collectFieldValues(item, new Map(schema.descriptors.map((descriptor) => [descriptor.identity, descriptor]))));
  const header = [...schema.topKeys, ...schema.descriptors.map((descriptor) => descriptor.column!)];
  const lines = includeHeader ? [header.map(quoteCsv).join(",")] : [];
  for (let index = 0; index < items.length; index++) {
    const item = items[index]!;
    const values = fieldValues[index]!;
    const row = [
      ...schema.topKeys.map((key) => formatCell(item[key], safety)),
      ...schema.descriptors.map((descriptor) => formatCell(values.get(descriptor.identity), safety)),
    ];
    lines.push(row.map(quoteCsv).join(","));
  }
  return lines.length === 0 ? "" : `${lines.join("\n")}\n`;
}

export function renderItemsCsv(items: JsonRecord[], safety: CsvSafety): string {
  if (items.length === 0) return "";

  const topKeys = [...new Set(items.flatMap((item) => Object.keys(item).filter((key) => key !== "fields")))].sort(compareText);
  const topKeySet = new Set(topKeys);
  const descriptors = new Map<string, FieldDescriptor>();
  const fieldValues = items.map((item) => collectFieldValues(item, descriptors));
  const ordered = Array.from(descriptors.values()).sort((left, right) => compareText(left.label, right.label) || compareText(left.identity, right.identity));

  let missingIndex = 0;
  for (const descriptor of ordered) {
    if (descriptor.key === undefined) descriptor.missingIndex = ++missingIndex;
  }

  const labelCounts = new Map<string, number>();
  for (const descriptor of ordered) labelCounts.set(descriptor.label, (labelCounts.get(descriptor.label) ?? 0) + 1);
  const usedColumns = new Set(topKeys);
  for (const descriptor of ordered) {
    const collides = topKeySet.has(descriptor.label) || (labelCounts.get(descriptor.label) ?? 0) > 1;
    const base = collides
      ? descriptor.key !== undefined
        ? `${descriptor.label} [${descriptor.key}]`
        : `${descriptor.label} [#${descriptor.missingIndex}]`
      : descriptor.label;
    descriptor.column = uniqueColumn(base, usedColumns);
    usedColumns.add(descriptor.column);
  }

  const header = [...topKeys, ...ordered.map((descriptor) => descriptor.column!)];
  const lines = [header.map(quoteCsv).join(",")];
  for (let index = 0; index < items.length; index++) {
    const item = items[index]!;
    const values = fieldValues[index]!;
    const row = [
      ...topKeys.map((key) => formatCell(item[key], safety)),
      ...ordered.map((descriptor) => formatCell(values.get(descriptor.identity), safety)),
    ];
    lines.push(row.map(quoteCsv).join(","));
  }
  return `${lines.join("\n")}\n`;
}

function collectFieldValues(item: JsonRecord, descriptors: Map<string, FieldDescriptor>): Map<string, unknown> {
  const values = new Map<string, unknown>();
  const missingOccurrences = new Map<string, number>();
  if (!Array.isArray(item["fields"])) return values;

  for (const value of item["fields"]) {
    if (value === null || typeof value !== "object" || Array.isArray(value)) continue;
    const field = value as JsonRecord;
    const key = typeof field["key"] === "string" && field["key"] !== "" ? field["key"] : undefined;
    const label = fieldLabel(field);
    if (label === "") continue;
    let identity: string;
    if (key !== undefined) {
      identity = `key:${key}`;
    } else {
      const occurrence = (missingOccurrences.get(label) ?? 0) + 1;
      missingOccurrences.set(label, occurrence);
      identity = `missing:${label}:${occurrence}`;
    }
    if (!descriptors.has(identity)) descriptors.set(identity, { identity, key, label });
    if (!values.has(identity)) values.set(identity, field["value"]);
  }
  return values;
}

function finalizeCsvSchema(topKeys: string[], descriptors: Map<string, FieldDescriptor>): CsvSchema {
  const ordered = Array.from(descriptors.values()).sort((left, right) => compareText(left.label, right.label) || compareText(left.identity, right.identity));
  let missingIndex = 0;
  for (const descriptor of ordered) {
    if (descriptor.key === undefined) descriptor.missingIndex = ++missingIndex;
  }

  const labelCounts = new Map<string, number>();
  for (const descriptor of ordered) labelCounts.set(descriptor.label, (labelCounts.get(descriptor.label) ?? 0) + 1);
  const usedColumns = new Set(topKeys);
  const topKeySet = new Set(topKeys);
  for (const descriptor of ordered) {
    const collides = topKeySet.has(descriptor.label) || (labelCounts.get(descriptor.label) ?? 0) > 1;
    const base = collides
      ? descriptor.key !== undefined
        ? `${descriptor.label} [${descriptor.key}]`
        : `${descriptor.label} [#${descriptor.missingIndex}]`
      : descriptor.label;
    descriptor.column = uniqueColumn(base, usedColumns);
    usedColumns.add(descriptor.column);
  }
  return { topKeys, descriptors: ordered };
}

function fieldLabel(field: JsonRecord): string {
  for (const key of ["name", "title", "key"] as const) {
    const value = field[key];
    if (typeof value === "string" && value !== "") return value;
  }
  return "";
}

function uniqueColumn(base: string, used: Set<string>): string {
  if (!used.has(base)) return base;
  for (let index = 1; ; index++) {
    const candidate = `${base} [#${index}]`;
    if (!used.has(candidate)) return candidate;
  }
}

function formatCell(value: unknown, safety: CsvSafety): string {
  if (value === null || value === undefined) return "";
  if (typeof value === "string") return safety === "spreadsheet" ? protectSpreadsheetString(value) : value;
  if (typeof value === "number") return Number.isFinite(value) ? canonicalJson(value) : "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  return canonicalJson(value);
}

function protectSpreadsheetString(value: string): string {
  if (value === "") return value;
  if (value[0] === "\t" || value[0] === "\r" || value[0] === "\n") return `'${value}`;
  const first = value.replace(/^[ \t]+/, "")[0];
  return first === "=" || first === "+" || first === "-" || first === "@" ? `'${value}` : value;
}

function canonicalJson(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "number") return JSON.stringify(Number.isFinite(value) ? value : null);
  if (typeof value === "boolean") return value ? "true" : "false";
  if (Array.isArray(value)) return `[${value.map((entry) => canonicalJson(entry === undefined ? null : entry)).join(",")}]`;
  if (typeof value === "object") {
    const record = value as JsonRecord;
    const entries = Object.keys(record)
      .filter((key) => record[key] !== undefined)
      .sort(compareText)
      .map((key) => `${JSON.stringify(key)}:${canonicalJson(record[key])}`);
    return `{${entries.join(",")}}`;
  }
  return "null";
}

function quoteCsv(value: string): string {
  return /[",\r\n]/.test(value) || /^\s/u.test(value) ? `"${value.replace(/"/g, '""')}"` : value;
}

function compareText(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

File: cli/internal/cli/csv.go
package cli

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

var jsonNumberPattern = regexp.MustCompile(`^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?$`)

type csvFieldDescriptor struct {
	identity     string
	key          string
	label        string
	missingIndex int
	column       string
}

type csvSchema struct {
	topKeys     []string
	descriptors []*csvFieldDescriptor
}

// buildCSVSchema freezes the header from the first page plus board field
// definitions. Later pages are rendered against this schema and cannot grow
// the retained header or force a full export pre-scan.
func buildCSVSchema(items []map[string]any, board any) csvSchema {
	topSet := map[string]struct{}{}
	descriptors := map[string]*csvFieldDescriptor{}
	if record, ok := board.(map[string]any); ok {
		if fields, ok := record["fields"].([]any); ok {
			for _, value := range fields {
				field, ok := value.(map[string]any)
				if !ok {
					continue
				}
				registerCSVDescriptor(field, descriptors)
			}
		}
	}
	for _, item := range items {
		for key := range item {
			if key != "fields" {
				topSet[key] = struct{}{}
			}
		}
		collectCSVFieldValues(item, descriptors)
	}
	return finalizeCSVSchema(topSet, descriptors)
}

func renderCSVHeader(schema csvSchema) ([]byte, error) {
	row := append(append([]string{}, schema.topKeys...), csvDescriptorColumns(schema.descriptors)...)
	return renderCSVRecord(row)
}

func renderCSVRow(item map[string]any, schema csvSchema, safety string) ([]byte, error) {
	descriptors := cloneCSVDescriptors(schema.descriptors)
	values := collectCSVFieldValues(item, descriptors)
	row := make([]string, 0, len(schema.topKeys)+len(schema.descriptors))
	for _, key := range schema.topKeys {
		value, err := formatCSVCell(item[key], safety)
		if err != nil {
			return nil, err
		}
		row = append(row, value)
	}
	for _, descriptor := range schema.descriptors {
		value, err := formatCSVCell(values[descriptor.identity], safety)
		if err != nil {
			return nil, err
		}
		row = append(row, value)
	}
	return renderCSVRecord(row)
}

func renderCSVRecord(row []string) ([]byte, error) {
	var buffer bytes.Buffer
	writer := csv.NewWriter(&buffer)
	if err := writer.Write(row); err != nil {
		return nil, err
	}
	writer.Flush()
	if err := writer.Error(); err != nil {
		return nil, err
	}
	return buffer.Bytes(), nil
}

func registerCSVDescriptor(field map[string]any, descriptors map[string]*csvFieldDescriptor) {
	key, _ := field["key"].(string)
	label := csvFieldLabel(field)
	if label == "" {
		return
	}
	identity := "key:" + key
	if key == "" {
		identity = "missing:" + label + ":" + strconv.Itoa(nextMissingCSVIndex(label, descriptors))
	}
	if _, exists := descriptors[identity]; !exists {
		descriptors[identity] = &csvFieldDescriptor{identity: identity, key: key, label: label}
	}
}

func nextMissingCSVIndex(label string, descriptors map[string]*csvFieldDescriptor) int {
	max := 0
	prefix := "missing:" + label + ":"
	for identity := range descriptors {
		if !strings.HasPrefix(identity, prefix) {
			continue
		}
		value, err := strconv.Atoi(strings.TrimPrefix(identity, prefix))
		if err == nil && value > max {
			max = value
		}
	}
	return max + 1
}

func finalizeCSVSchema(topSet map[string]struct{}, descriptorByIdentity map[string]*csvFieldDescriptor) csvSchema {
	topKeys := sortedSetKeys(topSet)
	descriptors := make([]*csvFieldDescriptor, 0, len(descriptorByIdentity))
	for _, descriptor := range descriptorByIdentity {
		descriptors = append(descriptors, descriptor)
	}
	sort.Slice(descriptors, func(i, j int) bool {
		if descriptors[i].label != descriptors[j].label {
			return descriptors[i].label < descriptors[j].label
		}
		return descriptors[i].identity < descriptors[j].identity
	})

	missingIndex := 0
	labelCounts := map[string]int{}
	for _, descriptor := range descriptors {
		labelCounts[descriptor.label]++
		if descriptor.key == "" {
			missingIndex++
			descriptor.missingIndex = missingIndex
		}
	}
	used := map[string]struct{}{}
	for _, key := range topKeys {
		used[key] = struct{}{}
	}
	for _, descriptor := range descriptors {
		base := descriptor.label
		_, topCollision := topSet[descriptor.label]
		if topCollision || labelCounts[descriptor.label] > 1 {
			if descriptor.key != "" {
				base = fmt.Sprintf("%s [%s]", descriptor.label, descriptor.key)
			} else {
				base = fmt.Sprintf("%s [#%d]", descriptor.label, descriptor.missingIndex)
			}
		}
		descriptor.column = uniqueCSVColumn(base, used)
		used[descriptor.column] = struct{}{}
	}
	return csvSchema{topKeys: topKeys, descriptors: descriptors}
}

func csvDescriptorColumns(descriptors []*csvFieldDescriptor) []string {
	columns := make([]string, 0, len(descriptors))
	for _, descriptor := range descriptors {
		columns = append(columns, descriptor.column)
	}
	return columns
}

func cloneCSVDescriptors(descriptors []*csvFieldDescriptor) map[string]*csvFieldDescriptor {
	cloned := make(map[string]*csvFieldDescriptor, len(descriptors))
	for _, descriptor := range descriptors {
		copy := *descriptor
		cloned[descriptor.identity] = &copy
	}
	return cloned
}

func renderItemsCSV(items []map[string]any, safety string) (string, error) {
	if len(items) == 0 {
		return "", nil
	}

	topSet := map[string]struct{}{}
	descriptorByIdentity := map[string]*csvFieldDescriptor{}
	fieldValues := make([]map[string]any, len(items))
	for index, item := range items {
		for key := range item {
			if key != "fields" {
				topSet[key] = struct{}{}
			}
		}
		fieldValues[index] = collectCSVFieldValues(item, descriptorByIdentity)
	}
	topKeys := sortedSetKeys(topSet)

	descriptors := make([]*csvFieldDescriptor, 0, len(descriptorByIdentity))
	for _, descriptor := range descriptorByIdentity {
		descriptors = append(descriptors, descriptor)
	}
	sort.Slice(descriptors, func(i, j int) bool {
		if descriptors[i].label != descriptors[j].label {
			return descriptors[i].label < descriptors[j].label
		}
		return descriptors[i].identity < descriptors[j].identity
	})

	missingIndex := 0
	labelCounts := map[string]int{}
	for _, descriptor := range descriptors {
		labelCounts[descriptor.label]++
		if descriptor.key == "" {
			missingIndex++
			descriptor.missingIndex = missingIndex
		}
	}
	used := map[string]struct{}{}
	for _, key := range topKeys {
		used[key] = struct{}{}
	}
	for _, descriptor := range descriptors {
		base := descriptor.label
		_, topCollision := topSet[descriptor.label]
		if topCollision || labelCounts[descriptor.label] > 1 {
			if descriptor.key != "" {
				base = fmt.Sprintf("%s [%s]", descriptor.label, descriptor.key)
			} else {
				base = fmt.Sprintf("%s [#%d]", descriptor.label, descriptor.missingIndex)
			}
		}
		descriptor.column = uniqueCSVColumn(base, used)
		used[descriptor.column] = struct{}{}
	}

	header := append([]string{}, topKeys...)
	for _, descriptor := range descriptors {
		header = append(header, descriptor.column)
	}
	var buffer bytes.Buffer
	writer := csv.NewWriter(&buffer)
	if err := writer.Write(header); err != nil {
		return "", err
	}
	for index, item := range items {
		row := make([]string, 0, len(header))
		for _, key := range topKeys {
			value, err := formatCSVCell(item[key], safety)
			if err != nil {
				return "", err
			}
			row = append(row, value)
		}
		for _, descriptor := range descriptors {
			value, err := formatCSVCell(fieldValues[index][descriptor.identity], safety)
			if err != nil {
				return "", err
			}
			row = append(row, value)
		}
		if err := writer.Write(row); err != nil {
			return "", err
		}
	}
	writer.Flush()
	if err := writer.Error(); err != nil {
		return "", err
	}
	return buffer.String(), nil
}

func collectCSVFieldValues(item map[string]any, descriptors map[string]*csvFieldDescriptor) map[string]any {
	values := map[string]any{}
	missingOccurrences := map[string]int{}
	fields, _ := item["fields"].([]any)
	for _, value := range fields {
		field, ok := value.(map[string]any)
		if !ok {
			continue
		}
		key, _ := field["key"].(string)
		label := csvFieldLabel(field)
		if label == "" {
			continue
		}
		identity := "key:" + key
		if key == "" {
			missingOccurrences[label]++
			identity = fmt.Sprintf("missing:%s:%d", label, missingOccurrences[label])
		}
		if _, exists := descriptors[identity]; !exists {
			descriptors[identity] = &csvFieldDescriptor{identity: identity, key: key, label: label}
		}
		if _, exists := values[identity]; !exists {
			values[identity] = field["value"]
		}
	}
	return values
}

func csvFieldLabel(field map[string]any) string {
	for _, key := range []string{"name", "title", "key"} {
		if value, ok := field[key].(string); ok && value != "" {
			return value
		}
	}
	return ""
}

func uniqueCSVColumn(base string, used map[string]struct{}) string {
	if _, exists := used[base]; !exists {
		return base
	}
	for index := 1; ; index++ {
		candidate := fmt.Sprintf("%s [#%d]", base, index)
		if _, exists := used[candidate]; !exists {
			return candidate
		}
	}
}

func formatCSVCell(value any, safety string) (string, error) {
	if value == nil {
		return "", nil
	}
	if text, ok := value.(string); ok {
		if safety == "spreadsheet" {
			text = protectSpreadsheetString(text)
		}
		return text, nil
	}
	return canonicalJSON(value)
}

func protectSpreadsheetString(value string) string {
	if value == "" {
		return value
	}
	if value[0] == '\t' || value[0] == '\r' || value[0] == '\n' {
		return "'" + value
	}
	trimmed := strings.TrimLeft(value, " \t")
	if trimmed != "" && strings.ContainsRune("=+-@", rune(trimmed[0])) {
		return "'" + value
	}
	return value
}

func canonicalJSON(value any) (string, error) {
	var buffer bytes.Buffer
	if err := writeCanonicalJSON(&buffer, value); err != nil {
		return "", err
	}
	return buffer.String(), nil
}

func writeCanonicalJSON(buffer *bytes.Buffer, value any) error {
	switch typed := value.(type) {
	case nil:
		buffer.WriteString("null")
	case string:
		encoded, _ := json.Marshal(typed)
		buffer.Write(encoded)
	case bool:
		if typed {
			buffer.WriteString("true")
		} else {
			buffer.WriteString("false")
		}
	case json.Number:
		if !jsonNumberPattern.MatchString(typed.String()) {
			return fmt.Errorf("invalid JSON number %q", typed)
		}
		buffer.WriteString(typed.String())
	case float64, float32, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
		encoded, err := json.Marshal(typed)
		if err != nil {
			return err
		}
		buffer.Write(encoded)
	case []any:
		buffer.WriteByte('[')
		for index, entry := range typed {
			if index > 0 {
				buffer.WriteByte(',')
			}
			if err := writeCanonicalJSON(buffer, entry); err != nil {
				return err
			}
		}
		buffer.WriteByte(']')
	case map[string]any:
		keys := make([]string, 0, len(typed))
		for key := range typed {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		buffer.WriteByte('{')
		for index, key := range keys {
			if index > 0 {
				buffer.WriteByte(',')
			}
			encoded, _ := json.Marshal(key)
			buffer.Write(encoded)
			buffer.WriteByte(':')
			if err := writeCanonicalJSON(buffer, typed[key]); err != nil {
				return err
			}
		}
		buffer.WriteByte('}')
	default:
		return fmt.Errorf("unsupported CSV JSON value %T", value)
	}
	return nil
}

func sortedSetKeys(set map[string]struct{}) []string {
	keys := make([]string, 0, len(set))
	for key := range set {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}
